lib/view_component/base.rb

```diff
@@ -5,6 +5,41 @@
 require "view_component/previewable"
 
 module ViewComponent
+  class CollectionBase
+    def initialize(component, options)
+      @component = component
+      @options = options
+    end
+
+    def render_in(view_context, &block)
+      as = as_variable(@options) || :item
+      args = @options.except(:items, :as)
+
+      @options[:items].map do |item|
```

Comment:
I don't think that is the optimal way to render large collections. I suggest creating a string buffer and concatenate each rendered component instead of creating an array, then joining it later. Look for some inspiration on  `ActionView::OutputBuffer` class and see how it works under the hood. :)

Reply:
Looks like `ActionView::OutputBuffer < ActiveSupport::SafeBuffer < String` (see [`buffers.rb`](https://github.com/rails/rails/blob/e2cf0b1d780b2e09f5270249ca021d94ce4fff9d/actionview/lib/action_view/buffers.rb#L21), and [`output_safety.rb`](https://github.com/rails/rails/blob/e2cf0b1d780b2e09f5270249ca021d94ce4fff9d/activesupport/lib/active_support/core_ext/string/output_safety.rb#L135)) and [`ActionView::RenderedCollection`](https://github.com/rails/rails/blob/e2cf0b1d780b2e09f5270249ca021d94ce4fff9d/actionview/lib/action_view/renderer/abstract_renderer.rb#L125) does `map/join` as well. We can definitely make this use a buffer, but is that optimal? Do we want to optimize for time? If so, a naive benchmark tells me that `join` is going to be faster.

```
  require "benchmark"
  def test_benchmark
    n = 1000000
    Benchmark.bmbm do |x|
      foo = [1] * n

      buffer = ActionView::OutputBuffer.new
      x.report("using ActionView::OutputBuffer") do
        foo.each do
          buffer << "foobar"
        end
        string = buffer.html_safe
      end

      x.report("using map and join") do
        string = foo.map do
          "foobar"
        end.join
      end
    end
  end
```

```
                                     user     system      total        real
using ActionView::OutputBuffer   0.449913   0.004264   0.454177 (  0.455215)
using map and join               0.085798   0.002497   0.088295 (  0.088758)
```

I'm new to rails internals, is there another builder like object besides `OutputBuffer`? I was also a little surprised that `render_in` had to return anything as I was expecting some internal buffer to be getting appended to as things are rendered. Is it possible that some of this is just for inline test rendering?

Reply:
@tclem lol. I thought it were using a string buffer instead of concatenating array items. did you benchmark appending to a simple string? We want to optimize for less string allocations and less memory usage.

Reply:
Your benchmark isn't really apples to apples because you're doing `html_safe` in one and not the other. I think we'd want to do a benchmark of a real case using the rendering of the view component instead of test strings. My gut says that Rails must be doing html_safe etc. inside the loop. 

Imagine that with rendering components it looks more like this - note the new html_safe in the join case:

```ruby
require "benchmark"
  def test_benchmark
    n = 1000000
    Benchmark.bmbm do |x|
      foo = [1] * n

      buffer = ActionView::OutputBuffer.new
      x.report("using ActionView::OutputBuffer") do
        foo.each do
          buffer << "foobar"
        end
        string = buffer.html_safe
      end

      x.report("using map and join") do
        string = foo.map do
          "foobar".html_safe
        end.join
      end
    end
  end
```

then

```
                                     user     system      total        real
using ActionView::OutputBuffer   0.373818   0.002911   0.376729 (  0.377495)
using map and join               0.821182   0.017310   0.838492 (  0.840472)
```

Reply:
Yup, you're right, both cases need to call `html_safe` (which is what's happening at the end of `render_in`), but I think they both need to call it inside the loop. The benchmark you pasted has the `OutputBuffer` case doing `html_safe` once at the end and the `map/join` case doing `html_safe` on each iteration. Let's make them the same.

>  I think we'd want to do a benchmark of a real case using the rendering of the view component instead of test strings. 

Love this suggestion! Here's an updated benchmark that actually calls `render_inline`:

``` ruby
require "benchmark"
def test_benchmark
  n = 10000
  @product = OpenStruct.new(title: "Bye")
  Benchmark.bmbm do |x|
    foo = [1] * n

    buffer = ActionView::OutputBuffer.new
    x.report("using ActionView::OutputBuffer") do
      foo.each do
        buffer << render_inline(ProductComponent.new(product: @product, extra: "abc"))
      end
    end

    x.report("using map and join") do
      foo.map do
        render_inline(ProductComponent.new(product: @product, extra: "abc"))
      end.join
    end
  end
end
```
```
                                     user     system      total        real
using ActionView::OutputBuffer   1.427219   0.023809   1.451028 (  1.461946)
using map and join               1.152875   0.026003   1.178878 (  1.204991)
```


I didn't dig too far, but array joins in ruby are [implemented in c](https://github.com/ruby/ruby/blob/5b287481befe03cc3e3dbc4b5571e21dbc523bae/array.c#L2365) and appear to be specialized for strings.

Reply:
Re-running benchmarks with the knowledge that map/join will require a final `html_safe`. The `OutputBuffer` version doesn't need this because it's already a safe buffer.

map/join still wins by a little.

``` ruby
require "benchmark"
def test_benchmark
  n = 10000
  @product = OpenStruct.new(name: "Soap")
  Benchmark.bmbm do |x|
    foo = [1] * n

    buffer = ActionView::OutputBuffer.new
    x.report("using ActionView::OutputBuffer") do
      foo.each do
        buffer << render_inline(ProductComponent.new(product: @product, notice: "abc"))
      end
    end

    x.report("using map and join") do
      foo.map do
        render_inline(ProductComponent.new(product: @product, notice: "abc"))
      end.join.html_safe
    end
  end
end
```

```
                                     user     system      total        real
using ActionView::OutputBuffer   1.543178   0.028958   1.572136 (  1.573645)
using map and join               1.285421   0.025860   1.311281 (  1.312286)
```